### PR TITLE
Generalise to `reddsa` crate

### PR DESCRIPTION
**Pull request by str4d**
An implementation of RedJubjub or RedPallas on its own is very small, and something that could be implemented separately for `zebrad` and `zcashd` ([although the Zcash Foundation has already proposed replacing `zcashd`'s implementation of RedJubjub with this one](https://github.com/zcash/librustzcash/pull/200)).

However, the introduction of FROST for threshold multisignatures adds significant implementation complexity. I worked out a rough breakdown of the codebase (module lines, including whitespace and comments):

- 32% keys, signatures, signing, single verification, common utilities / errors
- 27% batch verification
- 41% FROST

FROST in this crate is building on RedDSA, which was intentionally designed to be generic and curve-independent; this means that it should equally apply to RedJubjub (for Sapling) and RedPallas (for Orchard). But reimplementing all this logic for a second time in Rust is not an effective use of engineering resources (that could instead be directed to this implementation).

This PR takes the generalisation approach:

- The crate is renamed from `redjubjub` to `reddsa` (a crate name I have already secured for this purpose).
- `SpendAuth` and `Binding` traits are introduced. Both traits depend on the existing sealed traits (and use them to inherit common curve details).
  - The RedJubjub implementation is generalised to RedDSA. The RedJubjub-specific logic is moved into a `sapling` module.
  - The FROST implementation, which was previously restricted to RedJubjub SpendAuth signatures, is now similarly typed on `S: SpendAuth`.

Still todo:
- Rebase once #117 is merged.
- Write a new `redjubjub` crate that wraps `reddsa`. This would probably make sense to live in this repo as part of a workspace.
- Try to remove RedJubjub-specific code from `reddsa`. This might require un-sealing some traits, and then re-applying seals in `redjubjub`.

### CODE REVIEW

File: src/sapling.rs
Comment (str4d) [inline src/sapling.rs:13]: This is necessary because a `DefaultIsZeroes: Default` bound on `Foo<S>` requires `S: Default`. Per the comment in the commit message for https://github.com/ZcashFoundation/redjubjub/pull/63/commits/2745d378ab35e7bc3268e03d463c7b3c3f8e1cc7 we could replace `DefaultIsZeroes` with explicit `Zeroize` and `Drop` derives; I think we should do that (because having a panicking `Foo::default()` on an unconstructable type will create a Bad Time for someone.

File: README.md
Comment (daira) [inline README.md:3]: Update this and other references to RedJubjub in this file.

Comment (str4d) [reply README.md:3]: Yep, there is a bunch of this kind of cleanup work that we can do once there's agreement on the direction.

**Comment by str4d** [comment]
I've rebased this on #131 so it includes both the latest `ff` trait stack, and the latest changes from `main`.

The last time we discussed this PR, we decided it would make sense for it to implement a workspace with both `reddsa` and `redjubjub` crates. I have not implemented this yet.

**Comment by str4d** [comment]
To unblock an `orchard` release, @teor2345 and I decided to proceed with a separate minimal `reddsa` repo; we can decide on how to set up a workspace later. I've updated this PR to address @daira's comment. I will rebase the PR again once #131 and #137 are merged, and then this PR branch will become the `main` branch of the separate repo (to which I will also push Orchard-specific supporting code prior to publishing `reddsa 0.1.0`).

File: src/sapling.rs
Comment (conradoplg) [reply src/sapling.rs:13]: I created an issue to do this later https://github.com/ZcashFoundation/reddsa/issues/1

**Review by conradoplg — APPROVED** [review]
Looks good, thank you!

There are some minor readability issues which we can address later and I created issues for them, also for some TODOs you mentioned:

- https://github.com/ZcashFoundation/reddsa/issues/1
- https://github.com/ZcashFoundation/reddsa/issues/2
- https://github.com/ZcashFoundation/reddsa/issues/1

The idea here is to not merge this PR but move it to the `reddsa` repo, right?

**Comment by str4d** [comment]
> The idea here is to not merge this PR but move it to the `reddsa` repo, right?

Correct. Eventually we will likely rectify the two repositories into a single workspace, but to unblock the `orchard` crate the separate repository is faster. Once we're ready to create the workspace, we can open a new PR here with the changes from the `reddsa` repo, then delete the current `reddsa` repo and rename this repo to `reddsa`.

This PR is now https://github.com/ZcashFoundation/reddsa/commit/878dd1351b27c92446df6b22b4d87c4983017528, and that repo also has two extra commits with Orchard support: https://github.com/ZcashFoundation/reddsa/compare/878dd1351b27c92446df6b22b4d87c4983017528...b91559345c0ea4ed788d3d311996d5f8528c96b7

**Pull request closed without merging**
